crypto/pem/pem_passwd.c

```diff
@@ -0,0 +1,318 @@
+void openssl_console_release_mutex(void) {
+    CRYPTO_STATIC_MUTEX_unlock_write(&console_global_mutex);
+}
+
+int openssl_console_open(void) {
+    is_a_tty = 1;
+    assert(CRYPTO_STATIC_MUTEX_is_write_locked(&console_global_mutex));
+
+#if !defined(OPENSSL_WINDOWS)
+    if ((tty_in = fopen(DEV_TTY, "r")) == NULL) {
+        tty_in = stdin;
+    }
+    if ((tty_out = fopen(DEV_TTY, "w")) == NULL) {
+        tty_out = stderr;
+    }
+    if (TTY_get(fileno(tty_in), &tty_orig) == -1) {
+        if (errno == ENOTTY || errno == EINVAL || errno == ENXIO || errno == EIO
+          || errno == EPERM || errno == ENODEV) {
+            is_a_tty = 0;
+          } else {
+              OPENSSL_PUT_ERROR(PEM, ERR_R_INTERNAL_ERROR);
+              return 0;
+          }
+    }
+#else
+    DWORD console_mode;
+    HANDLE hStdIn = GetStdHandle(STD_INPUT_HANDLE);
+
+    // Check if console (equivalent to checking for /dev/tty on Linux)
+    if (GetConsoleMode(hStdIn, &console_mode)) {
+        // It's a real console, use conin$ and conout$ to bypass any redirection
+        if ((tty_in = fopen("conin$", "r")) == NULL) {
+            tty_in = stdin;
+        }
+        if ((tty_out = fopen("conout$", "w")) == NULL) {
+            tty_out = stderr;
+        }
+
+        tty_orig = console_mode;
+    } else {
+        // Not a console, use stdin/stderr
+        tty_in = stdin;
+        tty_out = stderr;
+        is_a_tty = 0;
+    }
+#endif
+    return 1;
+}
+
+int openssl_console_close(void) {
+    if (tty_in != stdin) {
+        fclose(tty_in);
+    }
+    if (tty_out != stderr) {
+        fclose(tty_out);
+    }
+
+    return 1;
+}
+
+static int openssl_console_echo_disable(void) {
+#if !defined(OPENSSL_WINDOWS)
+    OPENSSL_memcpy(&(tty_new), &(tty_orig), sizeof(tty_orig));
+    tty_new.TTY_FLAGS &= ~ECHO;
+
+    if (is_a_tty && (TTY_set(fileno(tty_in), &tty_new) == -1)) {
+        OPENSSL_PUT_ERROR(PEM, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+#else
+    if (is_a_tty) {
+        tty_new = tty_orig;
+        tty_new &= ~ENABLE_ECHO_INPUT;
+        SetConsoleMode(GetStdHandle(STD_INPUT_HANDLE), tty_new);
+    }
+#endif
+
+    return 1;
+}
+
+static int openssl_console_echo_enable(void) {
+#if !defined(OPENSSL_WINDOWS)
+    OPENSSL_memcpy(&(tty_new), &(tty_orig), sizeof(tty_orig));
+    if (is_a_tty && (TTY_set(fileno(tty_in), &tty_new) == -1)) {
+        OPENSSL_PUT_ERROR(PEM, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+#else
+    if (is_a_tty) {
+        tty_new = tty_orig;
+        SetConsoleMode(GetStdHandle(STD_INPUT_HANDLE), tty_new);
+    }
+#endif
+    return 1;
+}
+
+int openssl_console_write(const char *str) {
+    assert(CRYPTO_STATIC_MUTEX_is_write_locked(&console_global_mutex));
+    if (fputs(str, tty_out) < 0 || fflush(tty_out) != 0) {
+        OPENSSL_PUT_ERROR(PEM, PEM_R_PROBLEMS_GETTING_PASSWORD);
+        if (ferror(tty_out)) {
+            ERR_add_error_data(2, "System error: ", strerror(errno));
+        }
+        return 0;
```

Comment:
Here and other places we use `ferror`, we should call `clearerr` after to avoid it affecting subsequent attempts.